velox/functions/lib/TDigest.h

```diff
@@ -0,0 +1,483 @@
+  }
+}
+
+template <typename A>
+void TDigest<A>::mergeNewValues(
+    std::vector<int16_t>& positions,
+    double compression) {
+  if (numMerged_ < weights_.size()) {
+    VELOX_CHECK_LE(weights_.size(), std::numeric_limits<int16_t>::max());
+    positions.resize(weights_.size());
+    std::iota(positions.begin(), positions.end(), 0);
+    auto newBegin = positions.begin() + numMerged_;
+    auto compare = [this](auto i, auto j) { return means_[i] < means_[j]; };
+    if (!std::is_sorted(means_.begin() + numMerged_, means_.end())) {
+      std::sort(newBegin, positions.end(), compare);
+    }
+    std::inplace_merge(positions.begin(), newBegin, positions.end(), compare);
+    for (int i = 0; i < positions.size(); ++i) {
+      if (i == positions[i]) {
+        continue;
+      }
+      auto wi = weights_[i];
+      auto mi = means_[i];
+      auto j = i;
+      for (;;) {
+        auto k = positions[j];
+        if (k == i) {
+          break;
+        }
+        weights_[j] = weights_[k];
+        means_[j] = means_[k];
+        positions[j] = j;
+        j = k;
+      }
+      weights_[j] = wi;
+      means_[j] = mi;
+      positions[j] = j;
+    }
+    VELOX_DCHECK(std::is_sorted(means_.begin(), means_.end()));
+  }
+  merge(compression, weights_.data(), means_.data(), weights_.size());
+}
+
+template <typename A>
+void TDigest<A>::merge(
+    double compression,
+    const double* weights,
+    const double* means,
+    int count) {
+  VELOX_CHECK_GT(count, 0);
+  VELOX_CHECK_GE(weights_.size(), count);
+  if (reverseCompress_) {
+    // Run the merge in reverse every other merge to avoid left-to-right
+    // bias.
+    mergeImpl<true>(compression, weights, means, count);
+  } else {
+    mergeImpl<false>(compression, weights, means, count);
+  }
+  reverseCompress_ = !reverseCompress_;
+}
+
+template <typename A>
+template <bool kReverse>
+void TDigest<A>::mergeImpl(
+    double compression,
+    const double* weights,
+    const double* means,
+    int count) {
+  const auto totalWeight = std::accumulate(weights, weights + count, 0.0);
+  const auto invTotalWeight = 1 / totalWeight;
+  const auto normalizer =
+      (4 * std::log(totalWeight / compression) + 24) / compression;
+  auto maxSize = [normalizer](double q) { return q * (1 - q) * normalizer; };
+  double weightSoFar = 0;
+  numMerged_ = 0;
+  const int begin = kReverse ? count - 1 : 0;
+  auto notEnd = [&](auto i) INLINE_LAMBDA {
+    if constexpr (kReverse) {
+      return i >= 0;
+    } else {
+      return i < count;
+    }
+  };
+  constexpr int kStep = kReverse ? -1 : 1;
+  int j = kReverse ? count - 1 : 0;
+  weights_[j] = weights[begin];
+  means_[j] = means[begin];
+  for (int i = begin + kStep; notEnd(i); i += kStep) {
+    auto proposedWeight = weights_[j] + weights[i];
+    auto q0 = weightSoFar * invTotalWeight;
+    auto q2 = (weightSoFar + proposedWeight) * invTotalWeight;
+    if (proposedWeight <= totalWeight * std::min(maxSize(q0), maxSize(q2))) {
```

Comment:
I saw Presto-Java uses a normalizer here. Don't we need that too? https://github.com/prestodb/presto/blob/e416ff11482b92f812c2c47ef5ea0f0b19d85393/presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java#L325

Reply:
We have it captured in closure, see the definition of `maxSize`